src/osx/carbon/font.cpp

```diff
@@ -949,86 +969,151 @@ bool wxNativeFontInfo::FromString(const wxString& s)
+        m_encoding = (wxFontEncoding)l;
+        return true;
+    }
+    else if ( version == 2 )
+    {
+        token = tokenizer.GetNextToken();
+        if ( !token.ToLong(&l) )
+            return false;
+        bool underlined = l != 0;
 
-    return true;
+        token = tokenizer.GetNextToken();
+        if ( !token.ToLong(&l) )
+            return false;
+        bool strikethrough = l != 0;
+
+        token = tokenizer.GetNextToken();
+        if ( !token.ToLong(&l) )
+            return false;
+        wxFontEncoding encoding = (wxFontEncoding)l;
+
+        wxString xml = tokenizer.GetString();
+        xml+=GetPListPrefix();
+        wxCFStringRef plist(xml);
+        wxCFDataRef listData(CFStringCreateExternalRepresentation(kCFAllocatorDefault,plist,kCFStringEncodingUTF8,0));
+        wxCFDictionaryRef attributes((CFDictionaryRef) CFPropertyListCreateWithData(kCFAllocatorDefault, listData, 0, NULL, NULL));
+        CTFontDescriptorRef descriptor = NULL;
+        if (attributes != NULL)
+            descriptor = CTFontDescriptorCreateWithAttributes(attributes);
+        if (descriptor != NULL)
+        {
+            InitFromFontDescriptor(descriptor);
+            m_underlined = underlined;
+            m_strikethrough = strikethrough;
+            m_encoding = encoding;
+            return true;
+        }
+    }
+
+    return false;
 }
 
 wxString wxNativeFontInfo::ToString() const
 {
     wxString s;
 
-    s.Printf(wxT("%d;%s;%d;%d;%d;%d;%d;%s;%d"),
-        1, // version
-        wxString::FromCDouble(GetFractionalPointSize()),
-        GetFamily(),
-        (int)GetStyle(),
-        GetNumericWeight(),
-        GetUnderlined(),
-        GetStrikethrough(),
-        GetPostScriptName().GetData(),
-        (int)GetEncoding());
+    // version 2 is a streamed property list of the font descriptor as recommended by Apple
+    // prefixed by the attributes that are non-native to the native font ref like underline, strikethrough etc.
+    wxCFDictionaryRef attributes(CTFontDescriptorCopyAttributes(GetCTFontDescriptor()));
+
+    if (attributes != NULL)
+    {
+        CFPropertyListFormat format = kCFPropertyListXMLFormat_v1_0;
+        if (CFPropertyListIsValid(attributes, format))
+        {
+            wxCFDataRef listData(CFPropertyListCreateData(kCFAllocatorDefault, attributes, format, 0, NULL));
+            wxCFStringRef cfString( CFStringCreateFromExternalRepresentation( kCFAllocatorDefault, listData, kCFStringEncodingUTF8) );
+            wxString xml = cfString.AsString();
+            xml.Replace("\r",wxEmptyString,true);
+            xml.Replace("\t",wxEmptyString,true);
+            xml = xml.Mid(xml.Find("<plist"));
+
+            s.Printf("%d;%d;%d;%d;%s",
+                 2, // version
+                 GetUnderlined(),
+                 GetStrikethrough(),
+                 (int)GetEncoding(),
+                 xml);
+        }
+    }
+
+    if ( s.empty() )
+    {
+        // fallback to version 1
```

Comment:
Do you know if/when this can happen? `CTFontDescriptorCopyAttributes()` looks like it doesn't return NULL per docs. Nor is it clear to me how could the dictionary fail the `CFPropertyListIsValid()` test?

My concern here is that I think the serialization format revision should be bumped up to 2, with v1 remaining only reading-only backward compatibility code (which can be removed down the road when *N* years pass) instead of continuing to serialize in both v1 and v2 formats simultaneously.

Reply:
I've copied that code from the Apple Font docs in the serialization part

https://developer.apple.com/library/archive/documentation/StringsTextFonts/Conceptual/CoreText_Programming/FontOperations/FontOperations.html#//apple_ref/doc/uid/TP40005533-CH4-SW25

so I thought if they assume things could go wrong, I should do so too, and fall back in case of errors to what we know, so in practice this should not happen, but we'd be safe